bin/critterget.py

```diff
@@ -164,6 +174,181 @@ def apicall_with_response_code(uri, attribs=None):
+    """
+    Dispatches a saved search via Splunk's API and returns the results.
+
+    :param search_name: string, the name of the saved search
+    :param session_key: string, auth token for Splunk
+    :return:
+    """
+    saved_searches_uri = ('/servicesNS/admin/'
+                          'crittercism_integration/saved/searches')
+    search_jobs_uri = '/services/search/jobs/{}/results'
+    dispatch_url = None
+
+    all_saved_searches = splunk_api_call(saved_searches_uri, GET, session_key)
+
+    for entry in all_saved_searches[ENTRY]:
+        if entry[NAME] == search_name:
+            dispatch_url = entry[LINKS][DISPATCH]
+
+    if dispatch_url:
+        r = splunk_api_call(dispatch_url, POST, session_key)
+        sid = r[SID]
+        time.sleep(.25)  # This lets Splunk finish running the search
+    else:
+        print (u'{} MessageType="ApteligentError" Error: '
+               u'No saved searches returned by Splunk'.format(DATETIME_OF_RUN))
+        return
+
+    if sid:
+        search_url = search_jobs_uri.format(sid)
+        search_results = splunk_api_call(search_url, GET, session_key)
+    else:
+        print (u'{} MessageType="ApteligentError" Error: '
+               u'Could not dispatch search'.format(DATETIME_OF_RUN))
+        return
+
+    return search_results
+
+
+def get_last_run_time(session_key):
+    """
+    Retrieve the last time the connector ran from Splunk's API
+    :param session_key: string, auth token for Splunk
+    :return: datetime object, time of last run, or None
+    """
+    search_results = run_splunk_search(
+        'last_apps_call',
+        session_key
+    )
+
+    if not search_results or not search_results[RESULTS]:
+        print (u'{} MessageType="ApteligentError" Error: '
+               u'No search results returned by Splunk for {}'.format(
+                   'most recent run of the Apteligent connector',
+                   DATETIME_OF_RUN)
+              )
+        return
+
+    else:
+        last_run_string = search_results[RESULTS][0][UNDERSCORE_TIME]
+
+        return datetime.datetime.strptime(last_run_string,
+                                          '%Y-%m-%dT%H:%M:%S.%f+00:00')
+
+
+def what_to_run(session_key):
+    """
+    Based on the last time the connector ran, figure out which calls should
+    be made to the Apteligent API
+
+    TODO: This function is a stub that will be hooked to factories for
+    Apteligent API calls, once those are built.
+    :param session_key: string, auth token for Splunk
+    """
+    last_run = get_last_run_time(session_key)
+
+    if not last_run:
+        calls_to_run = ['basic calls', 'hour calls', 'daily calls']
+        print (u'{} MessageType="ApteligentTimestamp" LastRunTime="None" '
+               u'Running {}'.format(
+                   DATETIME_OF_RUN,
+                   calls_to_run)
+              )
+        return
+    else:
+        calls_to_run = ['basic calls']
+
+    floored_time = time_floor(last_run)
```

Comment:
This is more complicated than it looks -- it's actually asking two different questions. The first question is "is it time to run more than the usual ten-minute calls." We answer that by rounding off the minutes and seconds of the last run time, then checking whether that leaves us at the top of the hour (e.g. 15:00:00) or the start of a new day (00:00:00). The second question is "has the connector been running normally?" This is answered by checking whether the connector has run in the last hour and in the last day. If everything is running normally, the connector will run every ten minutes. If the hourly and daily calls are set to only run if the connector has been turned off, they will not be run regularly.